monitoring/nethealth.go

```diff
@@ -0,0 +1,411 @@
+// Implements health.Checker
+func (c *nethealthChecker) Check(ctx context.Context, reporter health.Reporter) {
+	probes, err := c.check(ctx)
+	if err != nil {
+		log.WithError(err).Error("Unable to verify nethealth.")
+	}
+
+	if probes.NumProbes() != 0 {
+		health.AddFrom(reporter, probes)
+		return
+	}
+	reporter.Add(NewSuccessProbe(c.Name()))
+}
+
+func (c *nethealthChecker) check(ctx context.Context) (health.Reporter, error) {
+	reporter := new(health.Probes)
+	pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		reporter.Add(NewProbeFromErr(c.Name(), "failed to retrieve k8s pods", err))
+		return reporter, trace.Wrap(err)
+	}
+
+	addr, err := c.getNethealthAddr(pods)
+	if err != nil {
+		return reporter, trace.Wrap(err, "failed to get local nethealth address")
+	}
+
+	mf, err := fetchMetrics(ctx, addr)
+	if err != nil {
+		reporter.Add(NewProbeFromErr(c.Name(), "unable to nethealth metrics", err))
+		return reporter, trace.Wrap(err, "failed to fetch metrics from %s", addr)
+	}
+
+	echoRequests, err := parseCounter(mf, echoRequestLabel)
+	if err != nil {
+		return reporter, trace.Wrap(err, "failed to parse echo requests")
+	}
+
+	echoTimeouts, err := parseCounter(mf, echoTimeoutLabel)
+	if err != nil {
+		return reporter, trace.Wrap(err, "failed to parse echo timeouts")
+	}
+
+	updated, err := c.updateStats(echoRequests, echoTimeouts)
+	if err != nil {
+		return reporter, trace.Wrap(err, "failed to update nethealth stats")
+	}
+
+	return c.verifyNethealth(updated)
+}
+
+// getNethealthAddr returns the address of the local nethealth pod.
+func (c *nethealthChecker) getNethealthAddr(pods *v1.PodList) (string, error) {
+	// Find nethealth pod with matching host ip address.
+	for _, pod := range pods.Items {
+		if pod.GetLabels()[nethealthLabel] != nethealthValue {
+			continue
+		}
+		if pod.Status.HostIP == c.AdvertiseIP {
+			return fmt.Sprintf("http://%s:%d", pod.Status.PodIP, c.NethealthPort), nil
+		}
+	}
+	return "", trace.NotFound("unable to find local nethealth pod")
+}
+
+// fetchMetrics collects the network metrics from the nethealth pod.
+func fetchMetrics(ctx context.Context, addr string) (metricFamilies map[string]*dto.MetricFamily, err error) {
+	client, err := roundtrip.NewClient(addr, "")
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to connect to nethealth service at %s.", addr)
+	}
+
+	resp, err := client.Get(ctx, client.Endpoint("metrics"), url.Values{})
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	var parser expfmt.TextParser
+	metricsFamilies, err := parser.TextToMetricFamilies(resp.Reader())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return metricsFamilies, nil
+}
+
+// updateStats updates netStats with new data collected in the provided
+// MetricFamily.
+func (c *nethealthChecker) updateStats(echoRequests, echoTimeouts map[string]float64) (updated []string, err error) {
+	if len(echoRequests) != len(echoTimeouts) {
```

Comment:
Do these lengths have to match? What do these maps map from->to?

Reply:
They should match. The metrics should come in a pair (requestTotal, timeoutTotal). I've just wrapped both into a `networkData` struct and have `parseMetrics` return `networkData`